synapse/rest/media/v1/media_storage.py

```diff
@@ -302,3 +316,44 @@ def write_to_consumer(self, consumer: IConsumer) -> Deferred:
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.open_file.close()
+
+
+class SpamMediaException(NotFoundError):
+    """The media was blocked by a spam checker, so we simply 404 the request (in
+    the same way as if it was quarantined).
+    """
+
+    pass
+
+
+@attr.s(slots=True)
+class ReadableFileWrapper:
+    """Wrapper that allows reading a file in chunks, yielding to the reactor,
+    and writing to an IO stream.
+
+    This is simplified `FileSender` that takes an IO object rather than an
+    `IConsumer`.
+    """
+
+    CHUNK_SIZE = 2 ** 14
+
+    clock = attr.ib(type=Clock)
+    path = attr.ib(type=str)
+
+    async def write_to_io(self, io: BinaryIO):
```

Comment:
Maybe this should just take a callback and if someone wants to write to a buffer they can?

Reply:
Ah I see this gives control of how often to request data to the spam checker while I'm suggesting keep that control, which essentially would be a `IProducer` and ask the spam checker to be an `IConsumer`?

Reply:
True